oauth_provider_jwt/models/oauth_provider_client.py

```diff
@@ -0,0 +1,187 @@
+
+        This check is only performed for asymetric algorithms
+        """
+        for client in self:
+            algorithm_prefix = client.jwt_algorithm[:2]
+            if algorithm_prefix in ALGORITHMS and client.jwt_private_key:
+                private_key = client._load_private_key()
+
+                if not isinstance(
+                   private_key, ALGORITHMS[algorithm_prefix]):
+                    raise exceptions.ValidationError(
+                        _('The private key doesn\'t fit the selected '
+                          'algorithm!'))
+
+    @api.multi
+    def generate_private_key(self):
+        """ Generate a private key for ECDSA and RSA algorithm clients """
+        for client in self:
+            algorithm_prefix = client.jwt_algorithm[:2]
+            if algorithm_prefix not in ALGORITHMS:
+                continue
+
+            if algorithm_prefix == 'ES':
+                key = ec.generate_private_key(
+                    curve=ec.SECT283R1,
+                    backend=default_backend(),
+                )
+            elif algorithm_prefix in ('RS', 'PS'):
+                key = rsa.generate_private_key(
+                    public_exponent=65537, key_size=2048,
+                    backend=default_backend(),
+                )
+
+            client.jwt_private_key = key.private_bytes(
+                encoding=Encoding.PEM,
+                format=PrivateFormat.TraditionalOpenSSL,
+                encryption_algorithm=NoEncryption(),
+            )
+
+    @api.multi
+    def _compute_jwt_public_key(self):
+        """ Compute the public key associated to the client's private key
+
+        This is only done for asymetric algorithms
+        """
+        for client in self:
+            if client.jwt_algorithm.startswith(('ES', 'RS', 'PS')) and \
+               client.jwt_private_key:
+                private_key = client._load_private_key()
+                client.jwt_public_key = private_key.public_key().public_bytes(
+                    Encoding.PEM, PublicFormat.SubjectPublicKeyInfo)
+            else:
+                client.jwt_public_key = False
+
+    @api.model
+    def _generate_jwt_payload(self, request):
+        """ Generate a payload containing data from the client """
+        data = {
+            'exp': datetime.utcnow() + timedelta(seconds=request.expires_in),
+            'nbf': datetime.utcnow(),
+            'iss': 'Odoo',
+            'aud': request.client.identifier,
+            'iat': datetime.utcnow(),
+            'user_id': request.client.generate_user_id(request.odoo_user),
+        }
+        if request.client.jwt_scope_id:
+            # Sudo as the token's user to execute the scope's filter with that
+            # user's rights
+            scope_data = request.client.jwt_scope_id.sudo(
+                user=request.odoo_user).get_data_for_model(
+                    'res.users', res_id=request.odoo_user.id)
```

Comment:
This should be broken into distinct operations:

```
scope = request.client.jwt_scope_id.sudo(
    user=request.odoo_user,
)
scope_date = scope.get_data_for_model(
    'res.users', res_id=request.odoo_user.id,
)
```

It's one more line, but way more clear IMO